src/timely-util/src/builder_async.rs

```diff
@@ -470,22 +510,24 @@ impl<G: Scope> OperatorBuilder<G> {
     {
         let operator_waker = self.operator_waker;
         let registered_wakers = self.registered_wakers;
-        let shared_frontiers = self.shared_frontiers;
-        let drain_pipe = self.drain_pipe;
+        let mut input_frontiers = self.input_frontiers;
+        let mut input_queues = self.input_queues;
         let mut output_flushes = self.output_flushes;
         let mut shutdown_handle = self.shutdown_handle;
         self.builder.build_reschedule(move |caps| {
             let mut logic_fut = Some(Box::pin(constructor(caps)));
             move |new_frontiers| {
-                // First, discover if there are any frontier notifications
-                {
-                    let mut old_frontiers = shared_frontiers.borrow_mut();
-                    for ((old, pending), new) in old_frontiers.iter_mut().zip(new_frontiers) {
-                        if PartialOrder::less_than(&old.borrow(), &new.frontier()) {
-                            *old = new.frontier().to_owned();
-                            *pending = true;
-                        }
+                for (i, queue) in input_queues.iter_mut().enumerate() {
+                    // First, discover if there are any frontier notifications
```

Comment:
Does this mean an operator will see frontier progress before seeing input data?

Reply:
Yes, but the frontier notification does not include the data being processed right afterwards. It is beneficial to serve the frontier notification first so that the maximum amount of work can be retired. This behavior is the same on `main`